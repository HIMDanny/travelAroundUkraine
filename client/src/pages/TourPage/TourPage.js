import {
  alpha,
  Box,
  Button,
  Container,
  Dialog,
  Link,
  Slide,
  Stack,
  styled,
  Typography,
  useMediaQuery,
} from '@mui/material';
import React, { useState, useEffect } from 'react';

import { useInView } from 'react-intersection-observer';

// React
import { useParams } from 'react-router-dom';

// Redux
import { useDispatch, useSelector } from 'react-redux';
import { fetchTour } from '../../store/slices/tourSlice/tourSlice';

import { TourAccordion, TourInfoDialog, TourReasonToChoose, ImageGallery } from '../../features/Tour/components';

const sections = [
  { title: 'About us', link: '#about-tour' },
  { title: 'Reasons to choose', link: '#reasons-to-choose' },
  { title: 'What is included?', link: '#included' },
];

const dates = { beginDate: new Date('2022-02-01'), endDate: new Date('2022-02-25') };

const HeaderContent = styled(Box)(({ theme }) => ({
  [theme.breakpoints.up('laptop')]: {
    borderBottom: `1px solid ${theme.palette.divider}`,
  },
}));

const Nav = styled('nav')(({ theme }) => ({
  paddingBlock: '15px',

  [theme.breakpoints.up('laptop')]: {
    paddingTop: '40px',
  },
}));

const NavLink = styled((props) => <Link variant="body1" underline="none" {...props} />)(({ theme }) => ({
  color: alpha(theme.palette.text.primary, 0.5),

  transition: theme.transitions.create('color'),

  '&:hover': {
    color: theme.palette.text.primary,
  },

  [theme.breakpoints.up('tablet')]: {
    fontSize: '16px',
  },
}));

const LinksWrapper = styled(Stack)(({ theme }) => ({
  flexDirection: 'row',
  gap: '20px',

  [theme.breakpoints.up('tablet')]: {
    justifyContent: 'center',
    gap: '40px',
  },
}));

const MainContent = styled(Box)(({ theme }) => ({
  paddingBlock: '10px',
  marginBottom: '150px',

  [theme.breakpoints.up('laptop')]: {
    paddingBlock: '40px',
  },
}));

const ContentWrapper = styled(Stack)({
  flexDirection: 'row',
  gap: '20px',
});

const Section = styled(Box)({
  scrollMarginTop: '20px',
});

const MobileDialogWrapper = styled(Box)(({ theme }) => ({
  paddingBlock: '10px',
  borderBlock: `1px solid ${theme.palette.divider}`,
}));

const Cost = styled(Typography)(({ theme }) => ({
  fontWeight: 700,
  marginBottom: 0,
  fontSize: 25,
  lineHeight: 1,
  color: theme.palette.primary.main,
}));

const FloatingDialog = styled(Box)(({ theme }) => ({
  position: 'fixed',
  width: '100%',
  left: 0,
  bottom: 0,
  borderTop: `1px solid ${theme.palette.divider}`,
  backgroundColor: '#fff',
  paddingBlock: '15px',
  zIndex: '1',
}));

const TourPage = () => {
  const matchesMediaQuery = useMediaQuery('(min-width: 1051px)');

  const [isOpen, setIsOpen] = useState(false);

  const [dialogRef, inView] = useInView({ root: null, rootMargin: '0px', threshold: 0.1 });

  const handleOpenDialog = () => setIsOpen(true);
  const handleCloseDialog = () => setIsOpen(false);

  const dispatch = useDispatch();
  const { itemNo } = useParams();
  // to be revised in future from re rendering and optimizing point of view, whether we pass needed data as props to components or useSelector directly in each component.
  const {
    imageUrls,
    name,
    description,
    reasons,
    professionalGuide,
    accommodation,
    meals,
    transferAlongTheRoute,
    travelInsurance,
    departs,
    duration,
    returns,
    currentPrice,
    region,
    categories,
    season,
    _id,
  } = useSelector((store) => store.tour.data);

  useEffect(() => {
    dispatch(fetchTour(itemNo));
  }, []);

  return (
    <>
      <HeaderContent>
        <Container>
          <Typography
            align="center"
            variant="h1"
            mt={17}
            mb={5}
            fontSize="50px"
            sx={{
              ':first-letter': {
                textTransform: 'capitalize',
              },
            }}
          >
            {name}
          </Typography>
          <Stack direction="row" justifyContent="center" alignItems="center" mb={3}>
            <Typography>{season} / </Typography>
            <Typography>{region} / </Typography>
            <Typography>{categories}</Typography>
          </Stack>
          <ImageGallery imageUrls={imageUrls} />
          <Nav>
            <LinksWrapper>
              {sections.map(({ title, link }) => (
                <NavLink key={title} href={link}>
                  {title}
                </NavLink>
              ))}
            </LinksWrapper>
          </Nav>
        </Container>
      </HeaderContent>

      <MainContent>
        <Container>
          <ContentWrapper>
            {matchesMediaQuery ? (
              <Box component="aside" sx={{ maxWidth: '370px', width: '100%' }}>
                <TourInfoDialog
                  dates={dates}
                  professionalGuide={professionalGuide}
                  accommodation={accommodation}
                  meals={meals}
                  transferAlongTheRoute={transferAlongTheRoute}
                  travelInsurance={travelInsurance}
                  departs={departs}
                  duration={duration}
                  returns={returns}
                  currentPrice={currentPrice}
                  id={_id}
                  itemNo={itemNo}
                />
              </Box>
            ) : null}

            <Box sx={{ flex: 1 }}>
              <Section id="about-tour">
                <Typography variant="h2">About tour</Typography>
                <Box sx={{ paddingBottom: 3 }}>
                  <Typography>{description}</Typography>
                </Box>
              </Section>

              <Section id="reasons-to-choose">
                <TourAccordion id="reasons-to-choose" title="Reasons to choose our tour">
                  <Stack direction="row" gap={4} flexWrap="wrap" mt={2} pl={2}>
                    {reasons
                      ? reasons.map((item, index) => (
                          <TourReasonToChoose description={item} number={index + 1} key={item} />
                        ))
                      : null}
                  </Stack>
                </TourAccordion>
              </Section>
            </Box>
          </ContentWrapper>

          {!matchesMediaQuery ? (
            <MobileDialogWrapper ref={dialogRef}>
              <Stack direction="row" justifyContent="space-between" alignItems="center">
                <Cost>€{currentPrice}</Cost>
                <Button sx={{ paddingInline: '30px' }} disableElevation onClick={handleOpenDialog}>
                  More info
                </Button>
              </Stack>
              <Dialog open={isOpen} onClose={handleCloseDialog} hideBackdrop fullScreen>
                <TourInfoDialog
                  dates={dates}
                  professionalGuide={professionalGuide}
                  accommodation={accommodation}
                  meals={meals}
                  transferAlongTheRoute={transferAlongTheRoute}
                  travelInsurance={travelInsurance}
                  departs={departs}
                  duration={duration}
                  returns={returns}
                  currentPrice={currentPrice}
                  id={_id}
                  itemNo={itemNo}
                  closeButton
                  handleClose={handleCloseDialog}
                />
              </Dialog>
            </MobileDialogWrapper>
          ) : null}
        </Container>

        {!matchesMediaQuery ? (
          <Slide in={!inView} direction="up" mountOnEnter unmountOnExit>
            <FloatingDialog>
              <Container>
                <Stack direction="row" justifyContent="space-between" alignItems="center">
                  <Cost>€{currentPrice}</Cost>
                  <Button sx={{ paddingInline: '30px' }} disableElevation onClick={handleOpenDialog}>
                    More info
                  </Button>
                </Stack>
              </Container>
            </FloatingDialog>
          </Slide>
        ) : null}
      </MainContent>
    </>
  );
};

export default TourPage;
