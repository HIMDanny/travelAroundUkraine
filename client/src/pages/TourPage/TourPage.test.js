import { render } from '@testing-library/react';
import TourPage from './TourPage';

import { mockAllIsIntersecting } from 'react-intersection-observer/test-utils';

jest.mock('@mui/material/Box', () => 'div');
jest.mock('@mui/material/Stack', () => ({ children }) => <div>{children}</div>);
jest.mock('@mui/material/Container', () => ({ children }) => <div>{children}</div>);
jest.mock('@mui/material/Typography', () => ({ children }) => <p>{children}</p>);
jest.mock('@mui/material/Button', () => ({ children }) => <button>{children}</button>);
jest.mock('@mui/material/Slide', () => ({ children }) => <div>{children}</div>);
jest.mock('@mui/material/Link', () => ({ children }) => <a>{children}</a>);
jest.mock('../../features/Tour/components/TourAccordion/TourAccordion', () => ({ children }) => <div>{children}</div>);
jest.mock('../../features/Tour/components/TourInfoDialog/TourInfoDialog', () => () => <div>TourInfoDialog</div>);
jest.mock('../../features/Tour/components/TourReasonToChoose/TourReasonToChoose', () => () => (
  <div>TourReasonToChoose</div>
));

let mockedMatchesMediaQuery = true;
jest.mock('@mui/material/useMediaQuery', () => jest.fn(() => mockedMatchesMediaQuery));

describe('Tour Page snapshot testing', () => {
  test('should Tour Page match snapshot', () => {
    const { asFragment } = render(<TourPage />);
    mockAllIsIntersecting(true);

    expect(asFragment()).toMatchSnapshot();
  });
});
