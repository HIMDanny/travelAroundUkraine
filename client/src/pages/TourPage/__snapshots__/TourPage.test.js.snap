// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tour Page snapshot testing should Tour Page match snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="c"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeMedium css-hh8ras-MuiButtonBase-root-MuiIconButton-root"
        id="p"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ArrowForwardIosIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6.23 20.23 8 22l10-10L8 2 6.23 3.77 14.46 12z"
          />
        </svg>
      </button>
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeMedium css-nl7jpl-MuiButtonBase-root-MuiIconButton-root"
        id="n"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ArrowForwardIosIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6.23 20.23 8 22l10-10L8 2 6.23 3.77 14.46 12z"
          />
        </svg>
      </button>
      <div
        class="swiper swiper-initialized swiper-horizontal swiper-pointer-events swiper-watch-progress mySwiper2"
      >
        <div
          class="swiper-wrapper"
          style="transition-duration: 0ms; transform: translate3d(NaNpx, 0px, 0px);"
        >
          <div
            class="swiper-slide swiper-slide-duplicate"
            data-swiper-slide-index="0"
          >
            <img
              alt=""
              src="http://placeimg.com/1200/800/nature"
            />
          </div>
          <div
            class="swiper-slide swiper-slide-duplicate"
            data-swiper-slide-index="1"
          >
            <img
              alt=""
              src="http://placeimg.com/800/1200/nature"
            />
          </div>
          <div
            class="swiper-slide swiper-slide-duplicate"
            data-swiper-slide-index="2"
          >
            <img
              alt=""
              src="http://placeimg.com/800/1200/nature"
            />
          </div>
          <div
            class="swiper-slide swiper-slide-duplicate"
            data-swiper-slide-index="3"
          >
            <img
              alt=""
              src="http://placeimg.com/1920/1080/nature"
            />
          </div>
          <div
            class="swiper-slide swiper-slide-duplicate"
            data-swiper-slide-index="4"
          >
            <img
              alt=""
              src="http://placeimg.com/1500/500/nature"
            />
          </div>
          <div
            class="swiper-slide swiper-slide-duplicate"
            data-swiper-slide-index="5"
          >
            <img
              alt=""
              src="http://placeimg.com/1500/500/nature"
            />
          </div>
          <div
            class="swiper-slide swiper-slide-duplicate"
            data-swiper-slide-index="6"
          >
            <img
              alt=""
              src="http://placeimg.com/800/1200/nature"
            />
          </div>
          <div
            class="swiper-slide swiper-slide-duplicate"
            data-swiper-slide-index="7"
          >
            <img
              alt=""
              src="http://placeimg.com/1500/500/nature"
            />
          </div>
          <div
            class="swiper-slide swiper-slide-duplicate"
            data-swiper-slide-index="8"
          >
            <img
              alt=""
              src="http://placeimg.com/1200/800/nature"
            />
          </div>
          <div
            class="swiper-slide swiper-slide-duplicate"
            data-swiper-slide-index="9"
          >
            <img
              alt=""
              src="http://placeimg.com/1200/800/nature"
            />
          </div>
          <div
            class="swiper-slide swiper-slide-duplicate"
            data-swiper-slide-index="10"
          >
            <img
              alt=""
              src="http://placeimg.com/1200/800/nature"
            />
          </div>
          <div
            class="swiper-slide"
            data-swiper-slide-index="0"
          >
            <img
              alt=""
              src="http://placeimg.com/1200/800/nature"
            />
          </div>
          <div
            class="swiper-slide"
            data-swiper-slide-index="1"
          >
            <img
              alt=""
              src="http://placeimg.com/800/1200/nature"
            />
          </div>
          <div
            class="swiper-slide"
            data-swiper-slide-index="2"
          >
            <img
              alt=""
              src="http://placeimg.com/800/1200/nature"
            />
          </div>
          <div
            class="swiper-slide"
            data-swiper-slide-index="3"
          >
            <img
              alt=""
              src="http://placeimg.com/1920/1080/nature"
            />
          </div>
          <div
            class="swiper-slide"
            data-swiper-slide-index="4"
          >
            <img
              alt=""
              src="http://placeimg.com/1500/500/nature"
            />
          </div>
          <div
            class="swiper-slide"
            data-swiper-slide-index="5"
          >
            <img
              alt=""
              src="http://placeimg.com/1500/500/nature"
            />
          </div>
          <div
            class="swiper-slide"
            data-swiper-slide-index="6"
          >
            <img
              alt=""
              src="http://placeimg.com/800/1200/nature"
            />
          </div>
          <div
            class="swiper-slide"
            data-swiper-slide-index="7"
          >
            <img
              alt=""
              src="http://placeimg.com/1500/500/nature"
            />
          </div>
          <div
            class="swiper-slide"
            data-swiper-slide-index="8"
          >
            <img
              alt=""
              src="http://placeimg.com/1200/800/nature"
            />
          </div>
          <div
            class="swiper-slide"
            data-swiper-slide-index="9"
          >
            <img
              alt=""
              src="http://placeimg.com/1200/800/nature"
            />
          </div>
          <div
            class="swiper-slide"
            data-swiper-slide-index="10"
          >
            <img
              alt=""
              src="http://placeimg.com/1200/800/nature"
            />
          </div>
          <div
            class="swiper-slide swiper-slide-duplicate"
            data-swiper-slide-index="0"
          >
            <img
              alt=""
              src="http://placeimg.com/1200/800/nature"
            />
          </div>
          <div
            class="swiper-slide swiper-slide-duplicate"
            data-swiper-slide-index="1"
          >
            <img
              alt=""
              src="http://placeimg.com/800/1200/nature"
            />
          </div>
          <div
            class="swiper-slide swiper-slide-duplicate"
            data-swiper-slide-index="2"
          >
            <img
              alt=""
              src="http://placeimg.com/800/1200/nature"
            />
          </div>
          <div
            class="swiper-slide swiper-slide-duplicate"
            data-swiper-slide-index="3"
          >
            <img
              alt=""
              src="http://placeimg.com/1920/1080/nature"
            />
          </div>
          <div
            class="swiper-slide swiper-slide-duplicate"
            data-swiper-slide-index="4"
          >
            <img
              alt=""
              src="http://placeimg.com/1500/500/nature"
            />
          </div>
          <div
            class="swiper-slide swiper-slide-duplicate"
            data-swiper-slide-index="5"
          >
            <img
              alt=""
              src="http://placeimg.com/1500/500/nature"
            />
          </div>
          <div
            class="swiper-slide swiper-slide-duplicate"
            data-swiper-slide-index="6"
          >
            <img
              alt=""
              src="http://placeimg.com/800/1200/nature"
            />
          </div>
          <div
            class="swiper-slide swiper-slide-duplicate"
            data-swiper-slide-index="7"
          >
            <img
              alt=""
              src="http://placeimg.com/1500/500/nature"
            />
          </div>
          <div
            class="swiper-slide swiper-slide-duplicate"
            data-swiper-slide-index="8"
          >
            <img
              alt=""
              src="http://placeimg.com/1200/800/nature"
            />
          </div>
          <div
            class="swiper-slide swiper-slide-duplicate"
            data-swiper-slide-index="9"
          >
            <img
              alt=""
              src="http://placeimg.com/1200/800/nature"
            />
          </div>
          <div
            class="swiper-slide swiper-slide-duplicate"
            data-swiper-slide-index="10"
          >
            <img
              alt=""
              src="http://placeimg.com/1200/800/nature"
            />
          </div>
        </div>
      </div>
      <div
        class="swiper swiper-initialized swiper-horizontal swiper-pointer-events swiper-watch-progress thumbsSwiper swiper-thumbs"
        style="height: 70px; width: auto;"
      >
        <div
          class="swiper-wrapper"
          style="transition-duration: 0ms; transform: translate3d(NaNpx, 0px, 0px);"
        >
          <div
            class="swiper-slide"
          >
            <img
              alt=""
              src="http://placeimg.com/1200/800/nature"
            />
          </div>
          <div
            class="swiper-slide"
          >
            <img
              alt=""
              src="http://placeimg.com/800/1200/nature"
            />
          </div>
          <div
            class="swiper-slide"
          >
            <img
              alt=""
              src="http://placeimg.com/800/1200/nature"
            />
          </div>
          <div
            class="swiper-slide"
          >
            <img
              alt=""
              src="http://placeimg.com/1920/1080/nature"
            />
          </div>
          <div
            class="swiper-slide"
          >
            <img
              alt=""
              src="http://placeimg.com/1500/500/nature"
            />
          </div>
          <div
            class="swiper-slide"
          >
            <img
              alt=""
              src="http://placeimg.com/1500/500/nature"
            />
          </div>
          <div
            class="swiper-slide"
          >
            <img
              alt=""
              src="http://placeimg.com/800/1200/nature"
            />
          </div>
          <div
            class="swiper-slide"
          >
            <img
              alt=""
              src="http://placeimg.com/1500/500/nature"
            />
          </div>
          <div
            class="swiper-slide"
          >
            <img
              alt=""
              src="http://placeimg.com/1200/800/nature"
            />
          </div>
          <div
            class="swiper-slide"
          >
            <img
              alt=""
              src="http://placeimg.com/1200/800/nature"
            />
          </div>
          <div
            class="swiper-slide"
          >
            <img
              alt=""
              src="http://placeimg.com/1200/800/nature"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="css-2k8nkm"
  >
    <div>
      <nav
        class="css-1ykz72r"
      >
        <div>
          <a>
            About us
          </a>
          <a>
            Reasons to choose
          </a>
          <a>
            What is included?
          </a>
        </div>
      </nav>
    </div>
  </div>
  <div
    class="css-1nmiu5j"
  >
    <div>
      <div>
        <div
          sx="[object Object]"
        >
          <div
            class="css-hyfiop"
            id="about-tour"
          >
            <p>
              About tour
            </p>
            <div
              sx="[object Object]"
            >
              <p>
                Each medieval city is full of various legends and mystical stories. Lviv is no exception because more than 760 years of history simply could not help but leave mysterious legends.
              </p>
              <p>
                The guide will tell and show that Lviv legends can impress no less than Lviv architecture. This tour will be filled with incredible stories - unusual stories from the life of the Galician capital.
              </p>
              <p>
                How can one explain the black colour of one of the houses on Rynok Square? How did Adam Senyavsky marry his daughter? How did the first bulletin board appear in Lviv? Why did one Lviv lady make a not entirely decent offer to the tram? Are the legends about Lviv ghosts true? What do Lviv courtyards hide?
              </p>
              <p>
                All this and more you can learn during this tour.
              </p>
              <p>
                Looks like today. The program of individual excursions: at any hour.
              </p>
            </div>
          </div>
          <div
            class="css-hyfiop"
            id="reasons-to-choose"
          >
            <div>
              <div>
                <div>
                  TourReasonToChoose
                </div>
                <div>
                  TourReasonToChoose
                </div>
                <div>
                  TourReasonToChoose
                </div>
              </div>
            </div>
          </div>
          <div
            class="css-hyfiop"
            id="included"
          >
            <div>
              <div>
                <div>
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-ci5apu-MuiSvgIcon-root"
                    data-testid="PersonIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                    />
                  </svg>
                  <p>
                    Professional guide
                  </p>
                </div>
                <div>
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-ci5apu-MuiSvgIcon-root"
                    data-testid="BedIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M21 10.78V8c0-1.65-1.35-3-3-3h-4c-.77 0-1.47.3-2 .78-.53-.48-1.23-.78-2-.78H6C4.35 5 3 6.35 3 8v2.78c-.61.55-1 1.34-1 2.22v6h2v-2h16v2h2v-6c0-.88-.39-1.67-1-2.22zM14 7h4c.55 0 1 .45 1 1v2h-6V8c0-.55.45-1 1-1zM5 8c0-.55.45-1 1-1h4c.55 0 1 .45 1 1v2H5V8z"
                    />
                  </svg>
                  <p>
                    Accomodation
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-13fe9vk"
      >
        <div>
          <p>
            €70
          </p>
          <button>
            More info
          </button>
        </div>
      </div>
    </div>
    <div>
      <div
        class="css-v912gw"
      >
        <div>
          <div>
            <p>
              €70
            </p>
            <button>
              More info
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
